Simplify adding extensions to `contao.image.valid_extensions`

zoglo opened the pull request:
### Description

Supersedes #6540 and implements a `+`/`-` syntax for the configuration based on the following two comments:

https://github.com/contao/contao/pull/6540#issuecomment-1814454318
https://github.com/contao/contao/pull/6540#issuecomment-1814410895

If we ever add a new defaults, the current approach can't be easily updated and people will have to change their configuration again by checking the defaults within the GitHub code.

### Usage:

**_Default behavior will stay the same_**
```yaml
contao:
    image:
        valid_extensions: ['jpg', 'jpeg', 'gif', 'png', 'tif', 'tiff', 'bmp', 'svg', 'svgz', 'webp', 'heic']
```

**_Adding an image format_**
```yaml
contao:
    image:
        valid_extensions: ['+heic']
```

**_Removing image formats_**
```yaml
contao:
    image:
        valid_extensions: ['-svg', '-svgz']
```

fritzmg commented:
I labeled the original issue as a bug as I thought it was an oversight that HEIC was not allowed by default. The +/- notation would be a feature though. wdyt?

zoglo commented:
You are right, this could and should be a feature.
Those who unlocked 'heic' in the past will have it enabled and this just makes it more convenient.

I started this long ago and shelved it - lemme change to 5.x

zoglo committed:
Add config information

zoglo committed:
Add configuration

zoglo committed:
Add missing test

zoglo committed:
Satisfy rector

zoglo committed:
Satisfy ECS

leofeyer commented on core-bundle/tests/DependencyInjection/ConfigurationTest.php:
```diff
@@ -54,6 +54,72 @@ public function testAddsTheImagineService(): void
+    public function testAddsImageExtension(): void
+    {
+        $extensions = ['jpg', 'jpeg', 'gif', 'png', 'tif', 'tiff', 'bmp', 'svg', 'svgz', 'webp'];
+        $extensions[] = 'heic';
```
```suggestion
        $extensions = ['jpg', 'jpeg', 'gif', 'png', 'tif', 'tiff', 'bmp', 'svg', 'svgz', 'webp', 'heic'];
```

zoglo replied:
Added it like that people would see the default values easier :)

leofeyer commented on core-bundle/tests/DependencyInjection/ConfigurationTest.php:
```diff
@@ -54,6 +54,72 @@ public function testAddsTheImagineService(): void
+    public function testRemovesImageExtension(): void
+    {
+        $extensions = ['jpg', 'jpeg', 'gif', 'png', 'tif', 'tiff', 'bmp', 'svg', 'svgz', 'webp'];
+        $extensions = array_diff($extensions, ['svg', 'svgz']);
```
```suggestion
        $extensions = ['jpg', 'jpeg', 'gif', 'png', 'tif', 'tiff', 'bmp', 'webp'];
```

zoglo replied:
Same as above

leofeyer commented on core-bundle/src/DependencyInjection/Configuration.php:
```diff
@@ -377,8 +377,50 @@ static function (array $value): array {
+                        ->ifTrue(
+                            static function (array $extensions): bool {
+                                foreach ($extensions as $extension) {
+                                    if (!preg_match('/^[+-]?.+$/', $extension)) {
```
```suggestion
                                    if (!preg_match('/^[+-]?[a-z0-9]+$/', $extension)) {
```

Would it make sense to restrict this further? Otherwise values like `\\\` would be considered valid, too.

leofeyer commented on core-bundle/src/DependencyInjection/Configuration.php:
```diff
@@ -377,8 +377,50 @@ static function (array $value): array {
+                                $newList = array_filter($extensions, static fn ($extension) => !\in_array($extension[0], ['-', '+'], true));
+
+                                if ([] !== $newList) {
+                                    return array_unique($newList);
```
Shouldn‘t `['+heic', '-svgz', 'jxl']` be turned into `['jxl', 'heic']` instead of returning just `['jxl']` here?

zoglo replied:
It's simply the example values that need an update since
```yaml
['+heic', '-svgz', 'jxl']
```
would always just return `['jxl']` as the only valid image extension, it's not a misconfiguration but doesn't make sense :)

`$newList` actually just removes every `+` and `-` and is an early return if a normal extension has been found.

Should we just update the example like this?
```php
->example(['+heic', '-svgz'])
```

leofeyer replied:
@ausi WDYT?

zoglo replied:
I updated it here but let's discuss what example should be best:
https://github.com/contao/contao/pull/7478/commits/15430cb8a89b22bd6987f067577a7a0e152b1abc

ausi replied:
We should use the same algorithm as we do here: https://github.com/contao/contao/blob/49c56bd953ec3b05bd91b45df9c835d9b374fb00/core-bundle/src/Intl/Locales.php#L151-L166

`['+heic', '-svgz', 'jxl']` should result in `['jxl', 'heic']` then.

It might make sense to move this algorithm to the `ArrayUtil` as we use it here too: https://github.com/contao/contao/blob/49c56bd953ec3b05bd91b45df9c835d9b374fb00/core-bundle/src/Intl/Countries.php#L77-L92

zoglo committed:
Simplify tests and update restriction for image sizes

* Apply suggestions from code review

Co-authored-by: Leo Feyer <1192057+leofeyer@users.noreply.github.com>

zoglo committed:
Update example

zoglo committed:
Reindex valid image extensions array

zoglo commented:
The failing checks are unrelated to this PR 🙈

leofeyer committed:
Merge branch '5.x' into fix/image-valid-extensions

zoglo committed:
Extract `filterValuesToIgnore` to ArrayUtil

zoglo committed:
Update tests for image extensions

zoglo committed:
Fix unrelated coding style

zoglo committed:
Remove `filterCountries` (now `ArrayUtils::filterValuesToIgnore`)

zoglo commented:
@ausi @leofeyer 
The newest changes should now follow the suggestion mentioned in https://github.com/contao/contao/pull/7478#discussion_r1733601924

I extracted the duplicate algorithm as well, I did find another occurence within [core-bundle/src/Doctrine/Backup/Config/AbstractConfig.php](https://github.com/contao/contao/pull/7478/files#diff-c3b66f0d61c14102544808484f2d519238543f3f028218a8f334813b9261de1d)

I also updated the tests according to the new style.

_Failing checks are unrelated_

ausi commented on core-bundle/contao/library/Contao/ArrayUtil.php:
```diff
@@ -149,4 +149,46 @@ public static function mapRecursive(callable $fn, array $arr): array
+	/**
+	 * Add, remove or replace values from the current array based on your configuration.
+	 */
+	public static function filterValuesToIgnore(array $current, array $filter, string|null $default = null): array
```
```suggestion
	/**
	 * Add, remove or replace values from the list based on your configuration.
	 */
	public static function alterListByConfig(array $list, array $config): array
```
Just a suggestion, WDYT @leofeyer ?

zoglo replied:
See 4a7201edead041bb70ceedbca8e6d60afdbd3f25

ausi commented on core-bundle/contao/library/Contao/ArrayUtil.php:
```diff
@@ -149,4 +149,46 @@ public static function mapRecursive(callable $fn, array $arr): array
+				unset($current[array_search($default, $current, true)]);
+			}
+			array_unshift($current, $default);
+		}
```
I think this is only needed in the `Locales` class and is a very special use case so I would only add it there and remove it from the `ArrayUtil`.

zoglo replied:
Unsure if it's a `special case` since it just makes sure a `default` is at the first place (Also just duplicate code 🤔)

Moving this function into ArrayUtils opens up new possibilities (not only for contao but for others), thus leaving this as an additional option may benefit us (it's not marked as `@internal`, is it 😊(All the ArrayUtils))

zoglo replied:
I reverted that one in 656dea7eb41479dcfcbce2871f0fd5cb7961e834

zoglo committed:
Revert filterLocales

zoglo committed:
Rename function to `alterListByConfig`

Co-authored-by: ausi <martin@auswoeger.com>

zoglo committed:
Use alterListByConfig within filterLocales

zoglo committed:
Merge branch '5.x' into fix/image-valid-extensions

ausi commented on core-bundle/contao/dca/tl_image_size.php:
```diff
@@ -328,9 +328,11 @@ public function getFormats(DataContainer|null $dc=null)
 		foreach ($this->getSupportedFormats() as $format => $isSupported)
 		{
-			if (!in_array($format, System::getContainer()->getParameter('contao.image.valid_extensions')))
+			if (!in_array($format, $imageExtensions))
```
This change seems to be unrelated to this pull request.

zoglo replied:
It's unrelated to this PR and was done almost 6 months ago (you could see it in the history).

Unsure about the performance but executing this for every supported format can be saved by just doing getting it once.

zoglo replied:
See: https://github.com/contao/contao/pull/7506

zoglo committed:
Revert `tl_image_size.php` change

leofeyer committed:
Merge branch '5.x' into fix/image-valid-extensions

leofeyer commented:
Thank you @zoglo.
